types.go

```diff
@@ -15,54 +15,62 @@ import (
 
 const defaultGradientAbsTol = 1e-6
 
-// EvaluationType is used by a Method to specify the objective-function
-// information needed at an x location.
-type EvaluationType uint
+// Operation represents the set of operations requested by Method at each
+// iteration. It is a bitmap of *Iteration and *Evaluation constants.
+type Operation uint64
 
-// Evaluation types can be composed together using the binary or operator, for
-// example 'FuncEvaluation | GradEvaluation' to evaluate both the function
-// value and the gradient.
+// Supported Operations. Individual operations must NOT be combined together by
+// the binary OR operator except for the various *Evaluations.
 const (
-	NoEvaluation   EvaluationType = 0
-	FuncEvaluation EvaluationType = 1 << (iota - 1)
+	// NoOperation does not specify any operation.
+	NoOperation Operation = 0
+	// InitIteration is sent to Recorder to indicate the initial location. All
+	// fields of the location to record must be valid.
+	// Methods must not return it.
+	InitIteration Operation = 1 << (iota - 1)
+	// MajorIteration indicates that a Method has found the next candidate
+	// location for an optimum and convergence should be checked.
+	MajorIteration
+	// PostIteration is sent to Recorder to indicate the final location reached
+	// during an optimization run. All fields of the location to record must be
+	// valid.
+	// Methods must not return it.
+	PostIteration
+	// FuncEvaluation is the request to evaluate the objective function.
```

Comment:
I changed it to "FuncEvaluation specifies that the objective function should be evaluated." I think using "should" is ok, because Problem.Status can return bad status, so the evaluation might not take place or Iterate() might not be called again.